### ISSUE TITLE
ZF3 + Doctrine2

### ISSUE THREAD

**Issue by tasselchof** [Codeception/Codeception#4181]
Hello!

Where is the problem using integration with Doctrine2 with ZF3? It's just not tested or there is known issues?

I am trying to use it and on first look it's working.

**Comment by Naktibalda** [comment]
https://github.com/Codeception/Codeception/blob/2.2.10/src/Codeception/Lib/Connector/ZF2.php#L160-L171
```php
if (!method_exists($serviceManager, 'addPeeringServiceManager')) {
    throw new ModuleException('Codeception\Module\ZF2', 'addServiceToContainer method is not compatible with ZF3');
}
```

Peering service manager was used to persist services between requests, but this functionality has been removed from service-manager 3.0, so another persistence mechanism is necessary.

**Comment by tasselchof** [comment]
@Naktibalda I will look deeper, but maybe you know - for what purpose peering was used?

Also I suppose, that not attaching it much better, than throwing an exception: this way it works.

**Comment by Naktibalda** [comment]
When test involves multiple requests ZF2 module restarts Zend\Application before each request to avoid polluting application state with data from earlier request.

Doctrine module wraps entire test in transaction and rolls it back after test to clean​ up. To be able to see the state of database Doctrine module must use the same database connection as Zend Application. To achieve that Doctrine entity manager created for the first request must be injected to the new instance of service manager before execution of later request.

**Comment by tasselchof** [comment]
Got it.

Am I right that overriding services (as it's done in [ZF unit tests](https://framework.zend.com/manual/2.4/en/tutorials/unittesting.html)) will do the trick?

We can override EntityManager on each test and it will be compatible with ZF2.

**Comment by Naktibalda** [comment]
Yes, that's right.

If you would like to work on it, I have some failing tests in https://github.com/Naktibalda/codeception-zf2-tests.

**Comment by tasselchof** [comment]
Yes, I will work on this issue.

**Comment by sleitz** [comment]
Are there any news on this? How hard or easy is it to make progress or help out?

**Comment by tasselchof** [comment]
Hi! I had no time to work on this, suppose to fix after 1-2 weeks. You have any progress?

**Comment by sleitz** [comment]
I have no idea where to start out. We have a lot of ZF expertise, but less with Doctrine and Codeception (only users there). So if you have a direction to go, we could try to help. Or wait :smile:

**Comment by bupy7** [comment]
Zend Framework 3 + Doctrine + Fixture https://github.com/bupy7/zf-app-blank/tree/master/module/ExCodeception/src

**Comment by Naktibalda** [comment]
You use the same instance of Application to server multiple requests, right?
It solves Doctrine issue, but causes some other issues with state persisting from one request to another, like #2964.

Though I don't know if ZF3 made any improvements in this area.

**Comment by bupy7** [comment]
> @Naktibalda You use the same instance of Application to server multiple requests, right?

Do you means HTTP-request?

**Comment by Naktibalda** [comment]
No, I mean Zend\Mvc\Application.

**Comment by bupy7** [comment]
@Naktibalda I destroy and create new instance of application for each request https://github.com/bupy7/zf-app-blank/blob/master/module/ExCodeception/src/Connector/ZF3.php#L51

May be I didn't understand you your position. Currently I have not a problem on this solution.

**Comment by bupy7** [comment]
Do you have some test for #2964?

**Comment by Naktibalda** [comment]
Oh, then you have the same problem as ZF2 module - when the test performs multiple requests, each instance of Application has a different instance of Doctrine EntityManager and later requests don't see changes made by earlier requests.
Also wrapping everything in transaction and rolling back it afterwards doesn't work.

I have a test for that here: https://github.com/Naktibalda/codeception-zf2-tests/blob/2.2/tests/doctrine/createTwoUsersAndGetListCept.php
https://github.com/Naktibalda/codeception-zf2-tests/blob/2.2/module/Application/src/Application/Controller/DoctrineController.php

**Comment by bupy7** [comment]
> you have the same problem as ZF2 module - when the test performs multiple requests

Yes, I agree. I do checks before I send next request. While I not need multiple request. I will test and think about this problem.

**Comment by bupy7** [comment]
@Naktibalda I think it is not bug. It is correct process. One request - one situation. After new request doing check past data is silly. Therefore, I think don't need fixing it. Need doing check after each request and not after everything ones.

**Comment by svycka** [comment]
I think I did it please test:
https://github.com/Codeception/Codeception/pull/5566
https://github.com/Naktibalda/codeception-zf2-tests/pull/10

**Comment by Naktibalda** [comment]
Fixed by #5566
